source/h1_stream.c

```diff
@@ -32,10 +33,94 @@ static void s_stream_update_window(struct aws_http_stream *stream, size_t increm
     aws_http_connection_update_window(stream->owning_connection, increment_size);
 }
 
+static int s_body_chunks_init(struct aws_h1_stream *stream) {
+    AWS_PRECONDITION(stream);
+    aws_linked_list_init(&stream->body_chunks.chunk_list);
+    stream->body_chunks.current_chunk = NULL;
+    stream->body_chunks.paused = false;
+    return aws_mutex_init(&stream->body_chunks.lock);
+}
+
+void s_clean_up_body_chunk(struct aws_http1_stream_chunk **chunk) {
+    AWS_PRECONDITION(chunk);
+    aws_h1_stream_release_chunk(*chunk);
+    *chunk = NULL;
+}
+
+void aws_h1_stream_body_chunks_clean_up(struct aws_h1_stream *stream) {
+    AWS_PRECONDITION(stream);
+    if (!stream->body_chunks.lock.initialized) {
+        return;
+    }
+    do {
+        if (NULL != stream->body_chunks.current_chunk) {
+            s_clean_up_body_chunk(&stream->body_chunks.current_chunk);
+        }
+    } while (aws_h1_populate_current_stream_chunk(&stream->body_chunks));
+    aws_mutex_clean_up(&stream->body_chunks.lock);
+}
+
+bool aws_h1_stream_is_paused(struct aws_h1_stream *stream) {
+    AWS_PRECONDITION(stream);
+    if (!stream->encoder_message.has_chunked_encoding_header) {
+        return false;
+    }
+    bool is_paused = false;
+    /* Begin critical section */
+    aws_h1_lock_chunked_list(&stream->body_chunks);
+    is_paused = stream->body_chunks.paused;
+    aws_h1_unlock_chunked_list(&stream->body_chunks);
+    /* End critical section */
+    return is_paused;
+}
+
+static int s_aws_h1_stream_write_chunk(struct aws_http_stream *stream_base, struct aws_http1_chunk_options *options) {
+    AWS_PRECONDITION(stream_base);
+    AWS_PRECONDITION(options);
+    struct aws_h1_stream *stream = AWS_CONTAINER_OF(stream_base, struct aws_h1_stream, base);
+    const size_t chunk_alloc_size = sizeof(struct aws_http1_stream_chunk);
+    struct aws_http1_stream_chunk *chunk = aws_mem_calloc(options->chunk_data->allocator, 1, chunk_alloc_size);
+    if (AWS_UNLIKELY(NULL == chunk)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_STREAM,
+            "static: Failed to initialize streamed chunk, error %d (%s).",
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+        return AWS_OP_ERR;
+    }
+    chunk->data = options->chunk_data;
+    chunk->data_size = options->chunk_data_size;
+    chunk->on_complete = options->on_complete;
+    chunk->user_data = options->user_data;
+    aws_linked_list_node_reset(&chunk->node);
+    if (AWS_OP_SUCCESS != aws_chunk_line_from_options(options, &chunk->chunk_line)) {
+        s_clean_up_body_chunk(&chunk);
+        return AWS_OP_ERR;
```

Comment:
set `on_complete` after calling aws_chunk_line_from_options().

That way, if something goes wrong with aws_chunk_line_from_options(), we're not firing the on-complete callback. Contract is: on_complete only fires if we successfully create the chunk

Reply:
yep, gotcha. my thought here was that is simplified the client handling because they could write one path to free resources, but this is cool too.